src/runtime2/loader/dom-ready.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  CanRender,
+  contextProp,
+  removeProp,
+  setProp,
+  subscribe,
+  unsubscribe,
+  useSyncEffect,
+} from '../../context';
+import {hasNextNodeInDocumentOrder} from '../../dom';
+
+/** @type {!ContextProp<?function(!Element)>} */
+// eslint-disable-next-line local/no-export-side-effect
+export const DomReadyProp = contextProp('DomReady', {
+  needsParent: true,
+  defaultValue: null,
+});
+
+/**
+ * @param {!Node} root
+ * @param {boolean} treeReady
+ */
+export function DomReadyProvider(root, treeReady) {
+  useSyncEffect(() => {
+    if (treeReady) {
+      return;
+    }
+
+    let pending = [];
+
+    const checkPending = () => {
+      for (let i = 0; i < pending.length; i++) {
+        const node = pending[i];
+        if (hasNextNodeInDocumentOrder(node, root)) {
+          // Remove resource before build to remove it from the pending list
+          // in either case the build succeed or throws an error.
+          pending.splice(i--, 1);
+          removeProp(node, CanRender, DomReadyProvider);
```

Comment:
This is actually important. The `CanRender` is computed as `AND()` over all inputs and parents. There can be many reasons why `CanRender` can be `false` and at least one of those would result in the actual "used" value computed as `false`. Thus the render blocker works this way: it first sets `CanRender` to `false` and then removes it once it can be unblocked.

One way to think about this is how this is usually done in CSS. e.g. for `display:none`. In CSS, we'd typically use several classes to control `display:none`. For instance, we might have `.not-build {display: none}`, `.hide-media-query {display:none}`, etc. When all such classes are removed the element finally becomes displayed. This is very similar how `CanRender` is used in this pull request.